tower-service/src/lib.rs

```diff
@@ -139,48 +137,84 @@ use std::task::{Context, Poll};
 ///     }
 /// }
 ///
+/// // The error returned if processing a request timed out
+/// #[derive(Debug)]
+/// pub struct Expired;
+///
+/// impl fmt::Display for Expired {
+///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         write!(f, "expired")
+///     }
+/// }
+///
+/// impl Error for Expired {}
+///
+/// // We can implement `Service` for `Timeout<T>` if `T` is a `Service`
 /// impl<T, Request> Service<Request> for Timeout<T>
 /// where
 ///     T: Service<Request>,
 ///     T::Future: 'static,
-///     T::Error: From<Expired> + 'static,
-///     T::Response: 'static
+///     T::Error: Into<Box<dyn Error>> + 'static,
+///     T::Response: 'static,
 /// {
+///     // `Timeout` doesn't modify the response type, so we use `T`'s response type
 ///     type Response = T::Response;
-///     type Error = T::Error;
+///     // Errors may be either `Expired` if the timeout expired, or the inner service's
+///     // `Error` type. Therefore, we return a boxed `dyn Error` trait object to erase
+///     // the error's type.
+///     type Error = Box<dyn Error>;
 ///     type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
 ///
 ///     fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+///         // Our timeout service is ready if the inner service is ready.
+///         // This is how backpressure can be propagated through a tree of nested services.
 ///        self.inner.poll_ready(cx).map_err(Into::into)
 ///     }
 ///
 ///     fn call(&mut self, req: Request) -> Self::Future {
-///         let timeout = tokio_timer::delay_for(self.timeout)
-///             .map(|_| Err(Self::Error::from(Expired)));
-///
-///         let fut = Box::pin(self.inner.call(req));
-///         let f = futures::select(fut, timeout)
-///             .map(|either| either.factor_first().0);
+///         // Create a future that completes after `self.timeout`
+///         let timeout = tokio::time::sleep(self.timeout);
```

Comment:
I wonder if we should use https://docs.rs/tokio/1.0.1/tokio/time/fn.timeout.html but I also kinda like that we show how to use `select!` and a `async move {}` instead of named types.

Reply:
I picked `sleep` because it matches how the current timeout middleware works but I'll let you decide which is more appropriate.